lib/segment/src/index/hnsw_index/graph_layers_builder.rs

```diff
@@ -365,29 +376,51 @@ impl GraphLayersBuilder {
 
                 for curr_level in (0..=linking_level).rev() {
                     let level_m = self.get_m(curr_level);
+                    let mut visited_list = self.get_visited_list_from_pool();
+
+                    visited_list.check_and_update_visited(level_entry.idx);
+
+                    let mut search_context = SearchContext::new(level_entry, self.ef_construct);
+
+                    self._search_on_level(
+                        &mut search_context,
+                        curr_level,
+                        &mut visited_list,
+                        &mut points_scorer,
+                    );
 
-                    let nearest_points = {
-                        self.search_existing_point_on_level(
-                            level_entry,
-                            curr_level,
-                            self.ef_construct,
-                            &mut points_scorer,
-                            point_id,
-                        )
-                    };
-
-                    if let Some(the_nearest) = nearest_points.iter().max() {
+                    if let Some(the_nearest) = search_context.nearest.iter().max() {
                         level_entry = *the_nearest;
                     }
 
                     let scorer = |a, b| points_scorer.score_internal(a, b);
 
                     if self.use_heuristic {
-                        let selected_nearest =
-                            Self::select_candidates_with_heuristic(nearest_points, level_m, scorer);
-                        self.links_layers[point_id as usize][curr_level]
-                            .write()
-                            .clone_from(&selected_nearest);
+                        let selected_nearest = {
+                            let mut existing_links =
+                                self.links_layers[point_id as usize][curr_level].write();
+                            {
+                                let ready_list = self.ready_list.read();
+                                for &existing_link in existing_links.iter() {
```

Comment:
This logic of existing links was moved from `search_existing_point_on_level`. You may copy this logic for non-heuristic case too